File: projet/client_lourd/src/app/services/room-validation.service/room-validation.service.ts
import { Injectable } from '@angular/core';
import { SocketClientService } from '@app/services/socket-client.service/socket-client.service';
import { ErrorDictionary } from '@common/browser-message/error-message/error-message';
// import { RoomValidationResult, UsernameValidation } from '@common/interfaces/socket-manager.interface';
// import { HOST_USERNAME } from '@common/names/host-username';
import {RoomValidationResult, UsernameValidation} from '@common/interfaces/socket-manager.interface';
import { SocketEvent } from '@common/socket-event-name/socket-event-name';
import {UsersService} from "@app/services/users.service/users.service";
import {firstValueFrom, Observable} from "rxjs";
import {User} from "@common/interfaces/user-data.interface";
import {TranslateService} from "@ngx-translate/core";

@Injectable({
    providedIn: 'root',
})
export class RoomValidationService {
    user$: Observable<User | null>;
    isActive: boolean = true;
    isLocked: boolean = false;
    isRoomIdValid: boolean = false;
    roomId: string | undefined = '';
    username: string | undefined;
    isUsernameValid: boolean = false

    constructor(private socketService: SocketClientService, private usersService: UsersService, private translate: TranslateService) {
        this.user$ = this.usersService.currentUserProfile$
        this.user$.subscribe((user: User | null) => {
            this.username = user?.uid;
        })
    }

    resetService() {
        this.isActive = true;
        this.isLocked = false;
        this.isRoomIdValid = false;
        this.roomId = '';
        this.username = '';
        this.isUsernameValid = false;
    }


    async verifyRoomId() {
        if (this.isOnlyDigit()) {
            return await this.sendRoomId();
        } else return this.translate.get('PLAYER_WAITING_PAGE.ROOM_CODE_PROMPT_ERRORS.VALIDATION_CODE_ERROR').toPromise();
    }

    async getCurrentUser() {
        return ((await firstValueFrom(this.usersService.currentUserProfile$)) as User);
    }

    async verifyUsername() {
        return await this.sendUsername();
    }

    async sendJoinRoomRequest() {
        const user = await this.getCurrentUser();
        return new Promise<string>((resolve) => {
            const usernameData = { roomId: Number(this.roomId), username: user.uid };
            this.socketService.send(SocketEvent.JOIN_GAME, usernameData, (isLocked: boolean) => {
                resolve(this.handleJoiningRoomValidation(isLocked));
            });
        });
    }

    private async sendUsername() {
        const user = await this.getCurrentUser();
        return new Promise<string>((resolve) => {
            const usernameData = { roomId: Number(this.roomId), username: user.uid };
            this.socketService.send(SocketEvent.VALIDATE_USERNAME, usernameData, async (data: UsernameValidation) => {
                resolve(await this.handleUsernameValidation(data));
            });
        });
    }

    private async sendRoomId() {
        return new Promise<string>((resolve) => {
            this.socketService.send(SocketEvent.VALIDATE_ROOM_ID, Number(this.roomId), async (data: RoomValidationResult) => {
                resolve(await this.handleRoomIdValidation(data));
            });
        });
    }

    private handleJoiningRoomValidation(isLocked: boolean) {
        this.isLocked = isLocked;
        return isLocked ? this.handleErrors(ErrorDictionary.ROOM_LOCKED) : '';
    }

    async handleUsernameValidation(data: UsernameValidation) {
        this.isUsernameValid = data.isValid;
        return data.isValid ? '' : this.translate.get('PLAYER_WAITING_PAGE.ROOM_CODE_PROMPT_ERRORS.BANNED_USER').toPromise(); // data.error is only if user is banned
    }

    async handleRoomIdValidation(data: RoomValidationResult) {
        let error = '';
        if (!data.isRoom) {
            error = this.handleErrors(await this.translate.get('PLAYER_WAITING_PAGE.ROOM_CODE_PROMPT_ERRORS.GAME_NOT_FOUND').toPromise());
        }
        else if (data.isLocked) error = this.handleErrors(await this.translate.get('PLAYER_WAITING_PAGE.ROOM_CODE_PROMPT_ERRORS.ROOM_LOCKED').toPromise());
        else this.isRoomIdValid = true;
        return error;
    }

    private handleErrors(errorType: string) {
        this.isRoomIdValid = false;
        this.isUsernameValid = false;
        return errorType;
    }

    private isOnlyDigit() {
        return this.roomId?.match('[0-9]{4}');
    }
}

File: projet/client_leger/polyquiz/lib/services/room_validation_service.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:rxdart/rxdart.dart';
import 'package:polyquiz/constants/socket-event.dart';
import 'package:polyquiz/models/user.dart';
import 'package:polyquiz/services/socket_service.dart';
import 'package:polyquiz/services/logged_in_user_service.dart';

class RoomValidationService with ChangeNotifier {
  final SocketService socketService;

  BehaviorSubject<User?> _userSubject = BehaviorSubject<User?>();
  Stream<User?> get user$ => _userSubject.stream;

  bool isActive = true;
  bool isLocked = false;
  bool isRoomIdValid = false;
  String? roomId = '';
  String? username;
  bool isUsernameValid = false;
  final LoggedInUserService loggedInUserService = LoggedInUserService.instance;
  User? userData;

  RoomValidationService({required this.socketService}) {
    this.loggedInUserService.reloadUser();
    this.userData = this.loggedInUserService.getUser();
  }

  void reloadUserData(){
    this.loggedInUserService.reloadUser();
    this.userData = this.loggedInUserService.getUser();
  }

  void resetService() {
    isActive = true;
    isLocked = false;
    isRoomIdValid = false;
    roomId = '';
    username = '';
    isUsernameValid = false;
    notifyListeners();
  }

  // Future<String> verifyRoomId() async {
  //   return isOnlyDigit() ? await sendRoomId() : 'VALIDATION_CODE_ERROR';
  // }

  Future<User> getCurrentUser() async {
    return (await this.userData) as User;
  }

  Future<String> verifyUsername() async {
    return await sendUsername();
  }

  // Future<String> sendJoinRoomRequest() async {
  //   final user = await getCurrentUser();
  //   return await Future<String>((resolve) {
  //     final usernameData = {'roomId': int.parse(roomId!), 'username': this.userData.uid};
  //     socketService.sendMessageWithAck(SocketEvent.JOIN_GAME, usernameData, (isLocked) {
  //       resolve(handleJoiningRoomValidation(isLocked));
  //     });
  //   });
  // }

  Future<String> sendUsername() async {
    final completer = Completer<String>();
    final user = await getCurrentUser();
    final usernameData = {'roomId': int.parse(roomId!), 'username': user.uid};
    socketService.sendMessageWithAck(
        SocketEvent.VALIDATE_USERNAME, usernameData, (data) {
      if (data != null) {
        completer.complete(handleUsernameValidation(data));
      } else {
        print('Failed to validate username');
        completer.completeError('Failed to validate username');
      }
    });
    return completer.future;
  }

  Future<Map<String, dynamic>> sendRoomId() async {
    final completer = Completer<Map<String, dynamic>>();
    socketService.sendMessageWithAck(SocketEvent.VALIDATE_ROOM_ID, int.parse(roomId!), (data) {
      print('I am here 2303');
      if (data != null) {
        print('I am here 030409');
        completer.complete(data);
        print(data);
      } else {
        print('Failed to validate roomID');
        completer.completeError('Failed to validate roomID');
      }
    });
    return completer.future;
  }

  String handleJoiningRoomValidation(bool isLocked) {
    this.isLocked = isLocked;
    return isLocked ? handleErrors('ROOM_LOCKED') : '';
  }

  String handleUsernameValidation(dynamic data) {
    isUsernameValid = data['isValid'];
    return data['isValid'] ? '' : data['error'];
  }

  String handleRoomIdValidation(dynamic data) {
    String error = '';
    if (!data['isRoom'])
      error = handleErrors('Le code ne correspond a aucune partie en cours. Veuillez réessayer');
    else if (data['isLocked'])
      error = handleErrors('"La partie est vérouillée. Veuillez réessayer."');
    else
      isRoomIdValid = true;
    return error;
  }

  String handleErrors(String errorType) {
    isRoomIdValid = false;
    isUsernameValid = false;
    return errorType;
  }

  bool isOnlyDigit() {
    return roomId?.contains(RegExp(r'^[0-9]{4}$')) ?? false;
  }
}
